Start here doc (#3449)
FusionAuth/fusionauth-site: astro/src/content/docs/get-started/start-here/index.mdx

---
title: "Getting Started with FusionAuth: A Beginner's Guide"
description: Learn how to set up FusionAuth, integrate authentication into your app, and implement advanced features like magic link logins in this developer-friendly guide.
section: get started
subcategory: start here
navOrder: 0

> codeRoot: https://raw.githubusercontent.com/synedra/fusionauth-first-steps-launch/main/

Author: Can we change this now that the changes to https://github.com/FusionAuth/fusionauth-example-express-start-here/ are merged (or approved to be merged)?

Reviewer: Done, committing in a moment.

---
import ListHostedLoginPagesUseCases from 'src/content/docs/_shared/_list-hosted-login-pages-use-cases.mdx';
import StartDiagram from 'src/diagrams/docs/get-started/start-here/launch-delegate-add.astro';
import Breadcrumb from 'src/components/Breadcrumb.astro';
import Aside from 'src/components/Aside.astro';
import InlineField from 'src/components/InlineField.astro';
import InlineUIElement from 'src/components/InlineUIElement.astro';
import DockerComposeFiles from 'src/content/docs/get-started/download-and-install/_docker-compose-files.mdx';
import { YouTube } from '@astro-community/astro-embed-youtube';

> import DockerSpinup from 'src/components/quickstarts/DockerSpinup.astro';

Author: is this the wrong component?

Reviewer: I am using the same component as the Quickstart, because it is correct in the get-started case as well.  I could make a copy and move that to get-started

To get started with FusionAuth, you need to do three things:

* Launch FusionAuth
* Delegate authentication
* Add functionality to your application

<StartDiagram alt="Launch, then delegate, then add functionality." />

Let's walk through an example where you'll add a Login With Google button to an application with no code changes required.

## Prerequisites

* git
* Docker (if you want to follow the Docker instructions)
* a GitHub account (if you choose to use GitPod instead)
* node (this was tested with `v22.12.0`

## Launch FusionAuth

FusionAuth is downloadable software, so you need to run it somewhere. For this guide, you can either:

* Use Docker to install and run FusionAuth and the application
* Use GitPod to run FusionAuth in the cloud in your GitPod account, with no software install needed.

Expand the section you want to follow and get started.

<details>
<summary>Docker</summary>
<DockerInstallAndStart codeRoot={frontmatter.codeRoot} kickstartUri={frontmatter.codeRoot + "/kickstart/kickstart.json"}/>
</details>

<details>
<summary>GitPod</summary>
<GitPodStart codeRoot={frontmatter.codeRoot} kickstartUri={frontmatter.codeRoot + "/kickstart/kickstart.json"}/>
</details>

> 

Author: There is no 'what did you just do' section for the launching of FusionAuth

Can we add that?

Reviewer: Added.

## Delegate Authentication

Once you have FusionAuth up and running, you need to set up your application or applications to delegate to FusionAuth. For this guide, there's an example 'Start Here' application which is already set up to do so. Let's look at that configuration and then start it up so you can log in.

> <details>

Author: What is the reason for using the accordions here? I understand using them with docker/gitpod because you pick one of these choices, but for everyone is going to go through all of these steps.

If we use these as components, do we lose the ability to use other components (like <Breadcrumb>)

Reviewer: Removed the extra accordions.

<summary>Open the Admin Interface</summary>

> <DelegateStep1 />

Author: If we are going to do these as components, can we use more meaningful names?

Reviewer: Updated the names and removed the extra cruft for accordions.

</details>

<DelegateStep2 />

> <details>

Author: Sometimes the details field is in the component, sometimes it is outside it? 

If we are going to use components, let's be consistent about what goes in them.

Reviewer: This is fixed.

<summary>Open the Application URL in your Browser and Login</summary>
<DelegateStep3 />
</details>

### What Did You Just Do?

You looked at some FusionAuth configuration, started a JavaScript application and checked out some cod, and logged in to the application using FusionAuth.

## Add Functionality

> 

Author: This section was missing from the gitpod installed instructions. Was that intentional, or is that a function of fusionauth.io not having this content live?

Reviewer: I added this for the launch section.

import StartDiagram from 'src/diagrams/docs/get-started/start-here/launch-delegate-add.astro';
import Breadcrumb from 'src/components/Breadcrumb.astro';
import Aside from 'src/components/Aside.astro';
import InlineField from 'src/components/InlineField.astro';
import InlineUIElement from 'src/components/InlineUIElement.astro';
import DockerComposeFiles from 'src/content/docs/get-started/download-and-install/_docker-compose-files.mdx';
import { YouTube } from '@astro-community/astro-embed-youtube';
import DockerSpinup from 'src/components/quickstarts/DockerSpinup.astro';
import DockerRequirements from 'src/components/quickstarts/DockerRequirements.astro'
import DockerInstallAndStart from 'src/components/quickstarts/DockerInstallAndStart.astro'
import GitPodStart from 'src/components/quickstarts/GitPodStart.astro'

To get started with FusionAuth, you need to do three things:

* Launch FusionAuth
* Delegate authentication
* Add functionality to your application

<StartDiagram alt="Launch, then delegate, then add functionality." />

Let's walk through an example where you'll add a Login With Google button to an application with no code changes required.

## Prerequisites

> * git

Author: Do we need node installed? Or should we have the sample app run in docker as well (in a node container)?

Reviewer: Oh, I think node can be required for Start Here, although adding the sample app as a docker container would be hella sweet.  Maybe a stretch goal :-)

* Docker (if you want to follow the Docker instructions)
* Github account (if you choose to use GitPod instead)

## Launch FusionAuth

FusionAuth is downloadable software, so you need to run it somewhere. For this guide, you can either

* Use Docker to install and run FusionAuth and the application
* Use GitPod to run FusionAuth in the cloud in your GitPod account, with no software install needed.

Expand the section you want to follow and get started.

<details>
<summary>Docker</summary>

<DockerRequirements />
<DockerInstallAndStart codeRoot={frontmatter.codeRoot} kickstartUri={frontmatter.codeRoot + "/kickstart/kickstart.json"}/>
</details>

<details>
<summary>GitPod</summary>

<GitPodStart codeRoot={frontmatter.codeRoot} kickstartUri={frontmatter.codeRoot + "/kickstart/kickstart.json"}/>

</details>

> ### Delegate Authentication

Author: ```suggestion
## Delegate Authentication
```

The next step is to delegate authentication for your application, and verify that the change is applied.  For this you will:
* Add the admin@example.com user to the application
* Logout from the administrative user
* Go to the test application
* Login as the administrative user

> ### Step 1. Bring up the administrative interface

Author: The headers should all have their first letter of each word capitalized (I believe that is in our style guide).

`Bring Up The Administrative Interface`

Reviewer: I think I got all of these.

In the previous step, you logged in to the system as an admin user.  Go back to that browser screen, which should be showing the administrative interface for FusionAuth.

> ### Step 2. Choose the correct user

Author: So this feels like a lot of steps. We could remove the 'add the registration' step and have it done in the quickstart.

Reviewer: Reworked the steps

* Navigate to "Users" in the left-hand sidebar (you may need to use the hamburger icon at the top left of the screen to see all options)
* Choose the "admin@example.com" user (Dinesh Chugtai)
* Choose "Manage" from the list of actions

> ### Step 3. Add registration to the application

Author: Is it really important to number the steps like this? I don't think we do that anywhere else.

Reviewer: Removed steps

* Scroll down to the bottom of the user page and click "Add registration"
* Choose the "Start Here" application (it may already be selected)
* Click the blue save icon at the top of the screen

### Step 4.  Log out from the administration interface

* If the left navigation bar is not visible, click the hamburger icon at the top left to see the options.
* At the top of the left navigation bar, click "Logout"

### Step 5. Startup the application

> In your shell, type the following to start up the application:

Author: Is this different between docker and gitpod?

Reviewer: I expanded this to show a Gitpod and Docker example.

```
cd app
npm install
npm run dev
```

> ### Step 6. Open the application

Author: ```suggestion
### Step 6. Open The Application URL In Your Browser
```

Reviewer: Changed to be more descriptive

> For gitpod:

Author: Should we have these as accordions for consistency?

Reviewer: Added accordions for Delegate steps

```
gp preview `gp url 8080`
```

If you are using docker, you can simply browse to http://localhost:8080

### Step 7. Check the login

* Login to the sample application using 'admin@example.com' and 'password'

> TODO maybe use nodejs quickstart as a base https://fusionauth.io/docs/quickstarts/quickstart-javascript-express-web

Author: ```suggestion
```

minimum to do here is:

* start the application (do those instructions differ between gitpod and docker? maybe)

> * show relevant sections for creating authorization url and exchanging tokens.

Author: When are we going to show the relevant sections of the codebase?

Author: Please use RemoteCode to pull it in from github rather than inlining it.

Reviewer: This is done.

![After entering users email address.](/img/docs/get-started/start-here/login-screen-magic-link-enter-email.png)

Now you can visit the [email inbox](http://localhost:1080) and you'll see your magic link email. 

![Email inbox with magic link message.](/img/docs/get-started/start-here/email-inbox-magic-link-message.png)

Click on the link, and you'll be logged into your application.

### What Did You Just Do?

You enabled Magic Link login for your application without touching a single line of code.

Congrats!

Let's talk about next steps.

## Quickstarts

Looking for an example in a particular language? Quickstarts walk you through the Launch and Delegate steps, leaving you with a running application in your framework or language of choice, to which you can add functionality.

If you have Docker installed and one of over 20 frameworks and languages, you can run through [our quickstarts](/docs/quickstarts).

Each quickstart includes a complete runnable application and a shell application you can build out by following the step-by-step integration instructions.

## Next Steps

> * Learn more about [other identity providers, including SAML, OIDC and Facebook](/docs/lifecycle/authenticate-users/identity-providers/).

Author: ```suggestion
* Set up [other identity providers, including SAML, OIDC and Facebook](/docs/lifecycle/authenticate-users/identity-providers/).
```

> * Learn about [passwordless options including magic links and passkeys](/docs/lifecycle/authenticate-users/passwordless/).

Author: ```suggestion
* Learn about other [passwordless options including passkeys](/docs/lifecycle/authenticate-users/passwordless/).
```

Reviewer: Changed.

* Enable [self-service registration](/docs/lifecycle/register-users/basic-registration-forms) to let users sign up for your application.
* Enable [multi-factor authentication (MFA)](/docs/lifecycle/authenticate-users/multi-factor-authentication). TOTP MFA is free, but other MFA methods require a paid license.

> * Sign up for a [trial FusionAuth Cloud instance](https://account.fusionauth.io/account/free-trial/start) which comes with a free Starter plan to explore [other paid features](/docs/get-started/core-concepts/premium-features).

Author: We may want to remove this, depending on free trial support (see https://inversoft.slack.com/archives/C051S8N8E/p1735592285725199 )

Reviewer: I'll ask Andy about this one

* **secure storage of user credentials and profile data**, wherever you want (you host or we host)
* **new functionality and bugfixes** from the security experts on the FusionAuth team, available to your users when you upgrade
* **the same experience** and application whether you develop locally, use FusionAuth in your CI/CD system, or run it in prod
* fine grained control of the version of FusionAuth you use; **upgrades happen on your schedule, not ours**

This guide will walk you through standing up a sample, pre-configured application that uses FusionAuth for login.

You'll also add the ability to log a user in using an emailed magic link.

## Overview of Using FusionAuth

To get started with FusionAuth, you need to do three things:

* Launch FusionAuth
* Delegate authentication
* Add functionality to your application

<StartDiagram alt="Launch, then delegate, then add functionality." />

Let's walk through an example where you'll add a Login With Google button to an application with no code changes required.

## Prerequisites

* git
* Docker (if you want to follow the Docker instructions)
* a GitHub account (if you choose to use GitPod instead)

> * node (this was tested with `v22.12.0`)

Reviewer: Fixed to the version I have

Once you have FusionAuth up and running, you need to set up your application or applications to delegate to FusionAuth. For this guide, there's an example `Start Here` application which is already set up to do so. Let's look at that configuration and then start it up so you can log in.

### Open the Admin Interface

* In the previous step, you logged in to the system as an admin user.  Go back to that browser screen, which should be showing the administrative interface for FusionAuth.
* Navigate to <InlineUIElement>Applications</InlineUIElement> in the left-hand sidebar. You may need to use the hamburger icon at the top left of the screen to see all options.
* Choose the `Start Here` application.
* Choose <InlineUIElement>Edit</InlineUIElement> from the <InlineUIElement>Actions</InlineUIElement> dropdown.

### Application Configuration

After the user signs in at FusionAuth, they need to get back to your application. In the FusionAuth application configuration you can view authorized redirect URLs and how tokens are generated/exchanged.

![The edit screen of the application.](/img/docs/get-started/start-here/edit-application.png)

### After the User Authenticates

> To understand how the JavaScript application integrates with FusionAuth, take a quick look at some of the code from the repository.

Author: ```suggestion
To understand how the JavaScript application integrates with FusionAuth, take a quick look at some of the code from the GitHub repository.
```

> We craft a URL that includes the `clientId` from the configuration you reviewed above, as well as some other parameters. This is where the user goes to log in with FusionAuth. TODO need to only show the `redirect(302, `fusionAuthURL/oauth2/authorize` line

Author: Need to update the code excerpt (the tags in the repo files) and remove the TODO here.

Reviewer: I'm not sure what you mean here, need more context.

Reviewer: Oh, I get it, you want a *different* code block which is smaller than what you have.  I will do that now.

Author: I took care of these, per our discussion.

<RemoteCode url="https://raw.githubusercontent.com/synedra/fusionauth-example-express-start-here/main/app/src/index.ts" lang="javascript" tags="login" />

> Once the user authenticates, the code at the configured redirect URL is called. This code sends parameters from the login event (the authorization code), the client identifier (clientId) and other information to get a valid token from FusionAuth. When the app receives this token, the user has fully been authenticated. TODO need to only show the try block here

Author: Need to update the code excerpt (the tags in the repo files) and remove the TODO here.

Reviewer: There is no "try" in the "login" tag.  Is it a different tag?

<RemoteCode url="https://raw.githubusercontent.com/synedra/fusionauth-example-express-start-here/main/app/src/index.ts" lang="javascript" tags="oauth-redirect" />

> Once you have the token, what should the app do with it? One option is to store it as a secure cookie. We do that here. TODO show cookie saving

Author: Need to update the code excerpt (the tags in the repo files) and remove the TODO here.

Reviewer: What does it mean to show cookie saving?  Is there specific code you want to show?  There are several different pieces of code surrounding the cookie stuff, or do you want to write something separate?

Author: I took care of this. hopefully it is decently clear in the revised doc.

We now can protect app functionality. For the account page, which requires authorization, the user's identity is checked and the browser is redirected to the main page if there is no valid token.

> <RemoteCode url="https://raw.githubusercontent.com/synedra/fusionauth-example-express-start-here/main/app/src/index.ts" lang="javascript" tags="account" />

Author: I notice these are still all synedra URLs. Is there a PR I need to approve so we can have these be updated?

Reviewer: Can we make it a separate PR?  Until the .gitpod stuff exists on the FA repository, everything will fail.

Author: This should be resolved by the fact this was merged, right? https://github.com/fusionauth/fusionauth-example-express-start-here/pull/2

### Open the Application URL in your Browser and Login

In your shell, type the following to start up the application:

<Code code="cd app && npm install && npm run dev" />

Point your browser to the application:

> GitPod:

Author: For consistency, should these be accordions?

Reviewer: Done

<Code code="gp preview `gp url 8080` --external" />

> Local:

Author: ```suggestion
Docker:
```

Author: This is also only going to work on a mac, which we didn't list in the prerequisites. 

Should we say something more widely applicable like "open your browser and navigate to `http://localhost:8080`?

Reviewer: That seems smart, I'll add it.

<Code code="open http://localhost:8080" />

> Now, login to the application in your browser using richard@example.com and 'password' to see the integration working.

Author: ```suggestion
Now, log in to the application in your browser:

* Click on the login button
* Authenticate using the email `richard@example.com` and the password `password`

You'll then be logged to the application.
```

### What Did You Just Do?

> You looked at some FusionAuth configuration, started a JavaScript application and checked out some cod, and logged in to the application using FusionAuth.

Author: ```suggestion
You looked at some FusionAuth configuration, started a JavaScript application and reviewed some code snippets, and logged in to the application using FusionAuth.
```

## Add Functionality

By delegating to FusionAuth, your application gets access to authentication and authorization functionality, such as magic link login, multi-factor authentication, social sign-on, self-service registration and more. This is where a lot of the value comes from.

> Enabling your users to log in using magic links can reduce friction and user frustration. Let's enable this on the application you just set up.

Author: ```suggestion
Enabling your users to log in using magic links can reduce friction and user frustration. Let's enable this on the "Start Here" application.
```

### Login With Magic Links

> To add a `Login With Magic Links` button, you usually have to configure an email server and then FusionAuth. But for this guide, we've already set up the email server. It is a locally running mock email server, and you can visit your inbox at [http://localhost:1080](http://localhost:1080).

Author: ```suggestion
To add a `Login With Magic Links` button, you usually have to get access to an email server and configure FusionAuth to use it. But for this guide, we've already set up the email server. It is a locally running mock email server, and you can visit your inbox at [http://localhost:1080](http://localhost:1080).
```

Author: Is that the right URL for gitpod? Probably not, right? Do we need to have two sets of instructions here?

Author: @synedra  any thoughts here?

> <Aside type="note">

Author: ```suggestion
```

> Typically you'd need to configure an email server and the email template, but this has been done for you in the setup steps above.

Author: ```suggestion
```

Author: Removing these because it is duplicative.

> </Aside>

Author: ```suggestion
```

> 

Author: ```suggestion

### Update FusionAuth Configuration

```

> Log out of your application and log in to the [Admin UI](http://localhost:9011/admin).

Author: ```suggestion
Log out of your application by clicking on the logout button. Then log in to the [Admin UI](http://localhost:9011/admin) to configure magic links.
```

> Navigate to <Breadcrumb>Applications</Breadcrumb> and then edit the ChangeBank application. Navigate to the <Breadcrumb>Security</Breadcrumb> tab and then the <Breadcrumb>Passwordless login</Breadcrumb> section.

Author: ```suggestion
Navigate to <Breadcrumb>Applications</Breadcrumb> and then edit the "Start Here" application. Navigate to the <Breadcrumb>Security</Breadcrumb> tab and then the <Breadcrumb>Passwordless login</Breadcrumb> section.
```

Click the <InlineField>Enable</InlineField> toggle.

![Enable the magic link login.](/img/docs/get-started/start-here/enable-magic-link-login.png)

Click the blue 'save' button to save your configuration change.

![Save the application.](/img/docs/get-started/start-here/save-application.png)

Now, log out of the admin UI.

> 

Author: ```suggestion

### Get A Magic Link And Log In

```

> [Visit your application](http://localhost:8080/) and click the <InlineUIElement>Login</InlineUIElement> button.

Author: Again, `localhost:8080` is not correct for GitPod, right?

![After entering users email address.](/img/docs/get-started/start-here/login-screen-magic-link-enter-email.png)

Now you can visit the [email inbox](http://localhost:1080) and you'll see your magic link email. 

![Email inbox with magic link message.](/img/docs/get-started/start-here/email-inbox-magic-link-message.png)

Click on the link, and you'll be logged into your application.

### What Did You Just Do?

You enabled Magic Link login for your application without touching a single line of code.

Congrats!

Let's talk about next steps.

## Quickstarts

Looking for an example in a particular language? Quickstarts walk you through the Launch and Delegate steps, leaving you with a running application in your framework or language of choice, to which you can add functionality.

If you have Docker installed and one of over 20 frameworks and languages, you can run through [our quickstarts](/docs/quickstarts).

Each quickstart includes a complete runnable application and a shell application you can build out by following the step-by-step integration instructions.

## Next Steps

> 

Author: ```suggestion

Now that you know a bit about FusionAuth, if you are interested in learning more about the authentication related functionality it enables, here's additional documentation.

```

### Get A Magic Link And Log In

> Visit your application.  To get to the application follow the instructions below.  Click on the arrow next to <InlineUIElement>Docker</InlineUIElement> or <InlineUIElement>GitPod</InlineUIElement> to see the instructions.

Author: ```suggestion
Visit your application. To get to the application follow the instructions below. Click on the arrow next to <InlineUIElement>Docker</InlineUIElement> or <InlineUIElement>GitPod</InlineUIElement> to see the instructions.
```
